### ISSUE TITLE
Delete Orphan File Attachments from Posts with PODS file Attach

### ISSUE THREAD

**Issue by dlinstedt** [pods-framework/pods#5724]
## Issue Overview
Enhancement Request
Subject: PODS file attachment management
Feature: During EDIT of a file attachment, when a file is "replaced" with a new upload / file, please delete the old attachment out of the uploads directory.

## Expected Behavior
Desired result:
WHEN a file (any file) in single file or multi-file list is "replaced/edited"
THEN delete old file in MEDIA/UPLOADS
and DELETE old POSTS entry (this is critical to keep wp_posts clean too)

Optionally: Add a checkbox to the settings of the Custom Post Type in the PODS creator, add the checkbox to the FILE upload type:
[ x ] delete old files when edited/replaced
and when turned on, fire that action as described above.

OR sell this as an add-on 

## Current Behavior
Result: check Media Folder -> old attachment still lives on
Check wp_posts -> old attachment still has PARENT ID = POST 1 and still exists in POSTS
check wp_postmeta -> Post meta for POST 1 now points to new attachment (correct behavior)

## Steps to Reproduce (for bugs)
Setup to see whats happening:
1) Create a new CPT
2) add a file attachment field (single file only)
3) save POD CPT
4) Create a new Post in the CPT
5) attach / upload File 1
6) Save/update
7) Edit Post 1
8) "EDIT" the file attachment (by replacing it with a new file upload)
9) Save/update

## Possible Solution
Install SNIPPETS, then install the PHP action hook code below

## WordPress Environment
wordpress 5.4.1
Elementor Pro 2.9.8
Astra Pro 2.5.0
PODS Pro 2.7.20

## Workaround or Alternate Solution Until Bug is Addressed
BELOW IS SHORTCODE THAT ACTUALLY DOES WHAT IS DESIRED ABOVE
EXCEPT: IT DOES NOT ADD A CHECKBOX TO THE OPTIONS PANEL

add_action('pods_api_post_save_pod_item', 'my_pods_post_save_function', 10, 3); 
 
function my_pods_post_save_function($pieces, $is_new_item, $id ) {
  global $wpdb;
	
        // contributed by: Dan Linstedt  https://danLinstedt.com
	// select all posts that ARE an ORPHAN file attachment (for this specific post)
	// where the PARENT is a PODS FILE ATTACH meta field
	// that have a count of ZERO where the FILE ATTACH ID
	// that was created by pods 
        // is NOT part of the CURRENT JSON attachment LIST
	$sql = "select ap.id,ap.guid
      from ".$wpdb->prefix."posts ap
      inner join ".$wpdb->prefix."postmeta apm
      on (ap.ID = apm.post_id)
      where
        ap.post_type = 'attachment'
        and apm.meta_key like '%file%'
        and ap.post_parent =".$id."
        and ap.post_parent in
        (select azm.post_id 
          from ".$wpdb->prefix."postmeta azm
          where azm.meta_key like '_pods%file'
          and azm.post_id = ap.post_parent
        )
        and 0 = 
        (select count(azz.meta_id) 
          from ".$wpdb->prefix."postmeta azz
          where azz.post_id = ap.post_parent and
            azz.meta_value like concat('%i:',ap.ID,';%')
        );";
 
	// calling wp_delete_attachment does the following:
	// delete POSTMETA related to this POST ID (all metadata)
	// delete POSTS related to this POST ID
	// delete FILE from the GUID
	$results = $wpdb->get_results($sql);
	foreach($results as $row) {
		$post_id = $row->id;
		wp_delete_attachment($post_id,true);
	}
}

**Comment by JoryHogeveen** [comment]
Hello @dlinstedt 

IMO, this is way too case specific. Since WordPress attachments are in fact just another post type that can be attached to anything it's difficult to determine exactly when to remove an attachment.

My view on this is to keep such a feature custom or in a separate addon.
@sc0ttkclark ??

**Comment by dlinstedt** [comment]
Hi Jory,

If you noticed, the code is specific to the single post-id that is being
saved.
So it does not affect *any* other attachments, except the ones for that
specific post where the post was attached to the file previously.
Again, the code is highly focused - to limit any damage to other posts, and
ignore all other attachments that are there.

This basically just keeps "garbage" from being uploaded and left all over
the media folder.

Thanks,
Dan

**Comment by JoryHogeveen** [comment]
I noticed that, but even then media attachments could have been used in more locations.
Even though a media attachment can only have one parent (the post it has been uploaded to) that doesn't mean it doesn't have more parents. This is a limitation of post types within WordPress.
I can use the same image in all my posts and still WordPress thinks only one of them is the parent.

It all depends on how you've created your site. I can imagine situations where this might be preferable but I can also imagine situations where this is not.

I fully understand your point of view but I need to view this in a larger perspective. That is why I think an addon or a snippet would be best. The user then has to have a more in depth understanding of what such a change means. Otherwise I'm sure our support channels will get new topics of users that lost their media files while using such a feature but not fully understand it.

**Comment by dlinstedt** [comment]
Agreed - and I can easily fix the SQL to accommodate / ignore attachments
that are part of more than one post, until they are "disassociated" with
all posts.

Even then, it would require an admin checkbox to "enable" the feature, and
documentation to explain it.

No worries, will keep it as a snippet for now.

**Comment by JoryHogeveen** [comment]
> Agreed - and I can easily fix the SQL to accommodate / ignore attachments
that are part of more than one post, until they are "disassociated" with
all posts.

Unfortunately you can't. If you add an image in your content this is done hardcoded, there isn't any form of association made.

Anyhow, could still be a nice addon option for some users. But I think such a feature should also make any attachments uploaded to such a field "exclusive" so they cannot be used in the regular media library. So in effect you'd have an internal library only available for that field.

**Comment by dlinstedt** [comment]
Alright you win.   However please note this:

1. Any file uploaded with PODS file attachment to begin with (to the pods
field) does receive (as you know) a POST attached by parent to the original
CPT
2. Therefore, any file (image or otherwise) previously managed by a PODS
CPT can actually be located using this method
3. Therefore, any file "still attached to the POST" but not found as
"active" in the JSON list (for current post or revisions) can be viewed as
disassociated and unused garbage.

UNLESS: PODS allows "attaching / associating" previously uploaded media -
as a "file attachment field" without ever creating a single sub-post
attachment type for it.
Then of course, all bets are off - and as you say, my thought process is
custom to purpose.

Anyhow, just thought I'd voice my opinion.

**Comment by JoryHogeveen** [comment]
Hi @dlinstedt 

It's not about winning ;) please don't get me wrong, I too was previously thinking about such a feature but came to the conclusion that the current way WordPress handles attachment makes this difficult to be fool/error proof.

Case 1:
- User uploads attachment in the library in another post.
- User then adds a new post and instead of uploading a new attachment in the field, just selects a previous one from the library
- The link between the second post is never made. So If this option is enabled the second post would lose it's attachment as well.

Case 2:
- A user adds a PDF file in a general post. Within the content of that field. Thie is not JSON, just plain HTML through the WordPress editor.
- User then adds a new post and again instead of uploading a new attachment in the field, just selects a previous one from the library
- If that second post removes it's attachment it will be deleted resulting in a dead PDF file link in the previous (first) post.

Then I'm not even talking about other plugins possibly using the same attachment file link under different metakeys etc. etc.

That is the reason I'm a bit careful with adding an option like this in Pods core until WordPress has a proper way of tracking unused attachments.
I do believe there are some third party plugins that do that but they encounter similar issues.

**Comment by dlinstedt** [comment]
Understood :)  Will consider these things, I was under the impression that
PODS meta file attach always uploaded new files, and didnt allow to choose
existing ones.  I think there might be a way, just have to consider it.
For now, my snippet will work for my purposes.

Thanks

**Comment by secondsky** [comment]
@JoryHogeveen  Is this snippet actually usable just as it is or does one need to change something to work with my own CPTs? Like changing something to the respective slug?

**Comment by JoryHogeveen** [comment]
Hi @secondsky 
Please read the whole thread. There is a reason such a feature isn't included in WordPress or Pods core.
Use at your own risk.

**Comment by secondsky** [comment]
@JoryHogeveen I am aware. In my case media is only attached to one CPT only.
My issue is that if I edit file or gallery through Pods (like delete a picture /exchange it), the file is still attached to the post (in media), just not visible in post edit in the pods fields. No clue why it stays attached to the post.
Is that maybe a new bug report worth? or is it connected to this topic?
Thx in advance for your help!! :-)

**Comment by JoryHogeveen** [comment]
Hi @secondsky 
This is simply how WordPress works. So long as there is no real mechanism for this we will not change this behavior.
No need to create a new bug report! It would be the same topic.

**Comment by secondsky** [comment]
@JoryHogeveen
Thx for the clarification! 
I understand that you dont want to change this.
So let me ask regarding this snipped shared here, is it 1:1 usable for any pods configuration? Or would i need to adapt something in it?
Sorry for asking so much ;-)
